### PR TITLE
ENH: enable skipna on groupby reduction ops

### LINKED ISSUE #15675: ENH: enable skipna on groupby reduction ops
Edit[rhshadach]: The following methods do not have a skipna argument in groupby, but do have such an argument on the Series/DataFrame variant.

- [ ] max
- [ ] mean
- [ ] median
- [ ] min
- [ ] prod
- [ ] sem
- [ ] std
- [ ] sum
- [ ] var

---

https://github.com/pandas-dev/pandas/issues/15674

```
In [19]: import pandas as pd
    ...: import numpy as np
    ...: d = {'l':  ['left', 'right', 'left', 'right', 'left', 'right'],
    ...:      'r': ['right', 'left', 'right', 'left', 'right', 'left'],
    ...:      'v': [-1, 1, -1, 1, -1, np.nan]}
    ...: df = pd.DataFrame(d)
    ...: 

In [20]: df.groupby('l').v.sum()
Out[20]: 
l
left    -3.0
right    2.0
Name: v, dtype: float64

In [21]: df.groupby('l').v.apply(lambda x: x.sum(skipna=False))
Out[21]: 
l
left    -3.0
right    NaN
Name: v, dtype: float64

```

ideally write [21] as
``df.groupby('l').v.sum(skipna=False)``

### PR DESCRIPTION
**Pull request by geoffrey-eisenbarth**
- [ ] closes #15675
- [ ] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

@mzeitlin11 Any help you can provide on what tests I should add, or where to start digging for similar `GroupBy` tests would be appreciated. As stated in PR #41399, I'm not super familiar with Cython, so hoping that the previous contributors (stale) PR is close to what's needed (I suppose the tests will tell us).

Thanks!

### CODE REVIEW

**Comment by mzeitlin11** [comment]
Thanks for reheating this @geoffrey-eisenbarth! Might not have much time this week, but will look deeper at the cython and testing when I can. For now, I'd recommend fixing up the easier issues like `precommit` first. 

With respect to testing, I'd take a look at the tests we have for groupby ops which do accept `skipna` (like `any/all`, the cumulative methods) and try to ensure similar coverage (as in make sure the new possibility `skipna=False` is covered for different types). Can look at tests in `pandas/tests/groupby` for inspiration. Some testing is nicely organized, some testing is stuffed in places like `pandas/tests/groupby/test_functions`.

Also, until we add tests, probably best to put the pr in draft state so others know it's not fully ready for review

File: pandas/_libs/groupby.pyx [hunk 1]
```diff
@@ -709,6 +723,11 @@ def group_mean(floating[:, ::1] out,
                     t = sumx[lab, j] + y
                     compensation[lab, j] = t - sumx[lab, j] - y
                     sumx[lab, j] = t
+                # don't skip nan
+                elif skipna == False:
```
Comment (mzeitlin11) [inline pandas/_libs/groupby.pyx:727]: This `skipna` is missing in the function arguments, causing a build failure

Comment (mzeitlin11) [reply pandas/_libs/groupby.pyx:727]: Also should be `elif not skipna`

File: pandas/_libs/groupby.pyx [hunk 2]
```diff
@@ -603,6 +612,10 @@ def group_prod(floating[:, ::1] out,
                 if val == val:
                     nobs[lab, j] += 1
                     prodx[lab, j] *= val
+                # don't skip nan
+                elif skipna == False:
```
Comment (mzeitlin11) [inline pandas/_libs/groupby.pyx:616]:
```suggestion
                elif not skipna:
```

File: pandas/_libs/groupby.pyx [hunk 3]
```diff
@@ -555,6 +559,10 @@ def group_add(add_t[:, ::1] out,
                         t = sumx[lab, j] + y
                         compensation[lab, j] = t - sumx[lab, j] - y
                         sumx[lab, j] = t
+                    # don't skip nan
+                    elif skipna == False:
```
Comment (mzeitlin11) [inline pandas/_libs/groupby.pyx:563]:
```suggestion
                    elif not skipna:
```

**Review by mzeitlin11 — COMMENTED** [review]
Just some initial syntax comments, will take a more thorough look at logic when I can

File: pandas/_libs/groupby.pyx [hunk 4]
```diff
@@ -530,6 +531,9 @@ def group_add(add_t[:, ::1] out,
                     else:
                         t = sumx[lab, j] + val
                     sumx[lab, j] = t
+                elif skipna == False:
+                    # NOTE: Does this case need to be considered?
```
Comment (jbrockmendel) [inline pandas/_libs/groupby.pyx:535]: yes.  if skipna is False and `not checknull(val)` (L524 above) then we `sumx[lab, j]` needs to be incremented by `val` (so will either become NaN or raise)

**Comment by geoffrey-eisenbarth** [comment]
Is CircleCI a new addition to the pandas contribution flow? I don't recall running into this when I worked on PR 41321. 

When I click on the details, it says I need a configuration specified for this project. Is that something I need to do on my own?

**Comment by mzeitlin11** [comment]
> Is CircleCI a new addition to the pandas contribution flow? I don't recall running into this when I worked on PR 41321.
> 
> When I click on the details, it says I need a configuration specified for this project. Is that something I need to do on my own?

If you merge master it should pick up the config and fix this failure

**Comment by jbrockmendel** [comment]
@geoffrey-eisenbarth planning to pick this back up?  it looked promising

**Comment by geoffrey-eisenbarth** [comment]
@jbrockmendel I was hoping to, since I rely on the old `sum(level=level, skipna=False)` syntax quite a bit. I know I let it fall behind and now two days ago was the birth my first child, so now it feels like I have zero time, but I will try to merge with master today to fix the CircleCI issues that came up (I'm not sure if CircleCI is new, but my previous contribution didn't seem to involve it) and see where that takes us.

**Comment by jbrockmendel** [comment]
No sweat, totally reasonable to have higher priorities under the circumstances.

Not sure about the circleci bit; often times I re-push and hope for CI issues to resolve themselves.

**Comment by jreback** [comment]
would likely take this, but needs to be passing an fully code-reviewed. closing as stale.

**Pull request closed without merging**